codec-http2/src/main/java/io/netty/handler/codec/http2/DelegatingDecompressorFrameListener.java

```diff
@@ -98,42 +97,37 @@ public int onDataRead(ChannelHandlerContext ctx, int streamId, ByteBuf data, int
                 // there is more data coming which will complete the decompression block. To allow for more data we
                 // return all bytes to the flow control window (so the peer can send more data).
                 decompressor.incrementDecompressedByes(compressedBytes);
-                processedBytes = compressedBytes;
-            } else {
-                try {
-                    decompressor.incrementDecompressedByes(padding);
-                    for (;;) {
-                        ByteBuf nextBuf = nextReadableBuf(channel);
-                        boolean decompressedEndOfStream = nextBuf == null && endOfStream;
-                        if (decompressedEndOfStream && channel.finish()) {
-                            nextBuf = nextReadableBuf(channel);
-                            decompressedEndOfStream = nextBuf == null;
-                        }
-
-                        decompressor.incrementDecompressedByes(buf.readableBytes());
-                        processedBytes += listener.onDataRead(ctx, streamId, buf, padding, decompressedEndOfStream);
-                        if (nextBuf == null) {
-                            break;
-                        }
-
-                        padding = 0; // Padding is only communicated once on the first iteration
-                        buf.release();
-                        buf = nextBuf;
+                return compressedBytes;
+            }
+            try {
+                int processedBytes = 0;
+                decompressor.incrementDecompressedByes(padding);
+                for (;;) {
+                    ByteBuf nextBuf = nextReadableBuf(channel);
+                    boolean decompressedEndOfStream = nextBuf == null && endOfStream;
+                    if (decompressedEndOfStream && channel.finish()) {
+                        nextBuf = nextReadableBuf(channel);
+                        decompressedEndOfStream = nextBuf == null;
+                    }
+
+                    decompressor.incrementDecompressedByes(buf.readableBytes());
+                    processedBytes += listener.onDataRead(ctx, streamId, buf, padding, decompressedEndOfStream);
```

Comment:
Good point ... we should consume immediately instead of returning the bytes. I'll add a comment on the return statement too.